Reopening of incorrectly merged `pad_sort_fix` correction.

sabogalc opened the pull request:
Previously, I corrected `pad_sort_fix` to correctly output the order of alphanumeric pads, like those found on BGA packages.
![Screenshot 2024-03-19 011926](https://github.com/whitequark/kicad-boardview/assets/53708281/de75469f-ef63-4644-acdb-54522bd56b95)
![Screenshot 2024-03-19 203507](https://github.com/whitequark/kicad-boardview/assets/53708281/ed250bd9-31ad-4f9c-a04e-ae3f844a0eba)

However, my solution hardcoded numbers, and you offered one that I mistakenly said worked since I did not refresh my KiCad plugin directory properly. The change resulted in this.
![image](https://github.com/whitequark/kicad-boardview/assets/53708281/3b439987-4ce3-4491-9438-8e1d294d9525)

I am confident there is still a better way to account for the numbering, but for now I just created this PR since the other one cannot be reopened. I apologize for all the mistakes and I'll be taking a break to wrap my head around Git before I come back.

sabogalc committed:
Add files via upload

Rewrote pad_sort_key to account for alphanumeric pads.

sabogalc committed:
Add files via upload

sabogalc committed:
Update pcbnew2boardview.py

Co-authored-by: Catherine <whitequark@whitequark.org>

sabogalc committed:
Update pcbnew2boardview.py

sabogalc committed:
Add files via upload

whitequark reviewed (commented):
Could you test the code in https://github.com/whitequark/kicad-boardview/pull/17#discussion_r1532876796 please?

sabogalc committed:
Add files via upload

Implemented natural_sort_key fix from https://github.com/whitequark/kicad-boardview/pull/17#discussion_r1532876796

sabogalc commented:
> Could you test the code in [#17 (comment)](https://github.com/whitequark/kicad-boardview/pull/17#discussion_r1532876796) please?

That code works and I have committed it to this branch (I double checked that the plugins were refreshed).

There is a slight discrepancy however. My revised `pad_sort_key` put pins with blank pin numbers at the end of the sort, so that something like a NPTH would not be stored as pin 1, as seen below.
![image](https://github.com/whitequark/kicad-boardview/assets/53708281/2a3d18cf-f33b-4c61-95e1-a035c999e536)
The `natural_sort_key` function does not account for this, and allows for no-connects and NPTH pins to be pin 1.
![image](https://github.com/whitequark/kicad-boardview/assets/53708281/62d4a180-0b6e-4076-b97b-6a8139ca2662)

whitequark commented:
There's now a merge conflict in this branch for some reason.

whitequark commented:
> The `natural_sort_key` function does not account for this, and allows for no-connects and NPTH pins to be pin 1.

Ah yeah I forgot about that bit, sorry

sabogalc committed:
Merge branch 'main' into padSortFix

sabogalc commented:
> > The `natural_sort_key` function does not account for this, and allows for no-connects and NPTH pins to be pin 1.
> 
> Ah yeah I forgot about that bit, sorry

What do you think of

```
def natural_sort_key(s):
    is_blank = s.strip() == ''
    return (is_blank, [int(text) if text.isdigit() else text.casefold()
                       for text in re.compile('([0-9]+)').split(s)])
```

I know I do that thing with a bunch of commits and I'm kind of doing it here too so I am trying to avoid that.

whitequark commented:
Yep that works

sabogalc committed:
Account for blank pin names

This commit places blank pin names (NCs, NPTH, etc.) at the end of the sort so that they are not mistakenly pin 1.

sabogalc commented:
Okay I've committed it with GitHub Desktop (which is hopefully more correct) and I think it should be good to go now.
